src/Compilers/Core/Portable/RealParser.cs

```diff
@@ -358,14 +342,24 @@ public static DecimalFloatingPointString FromSource(string source)
                     int firstExponent = i;
                     int lastExponent = i;
                     while (i < source.Length && source[i] >= '0' && source[i] <= '9') lastExponent = ++i;
-                    int exponentMagnitude = int.Parse(source.Substring(firstExponent, lastExponent - firstExponent));
-                    if (exponentSign == '-')
+
+                    int exponentMagnitude = 0;
+
+                    if (int.TryParse(source.Substring(firstExponent, lastExponent - firstExponent), out exponentMagnitude) &&
+                        exponentMagnitude <= MAX_EXP_QUAD)
                     {
-                        exponent -= exponentMagnitude;
+                        if (exponentSign == '-')
+                        {
+                            exponent -= exponentMagnitude;
+                        }
+                        else
+                        {
+                            exponent += exponentMagnitude;
```

Comment:
This addition can overflow (when the sum of `exponent` and `exponentMagnitude` does not fit into an int). That is not handled correctly here. Similarly for the subtraction in the other branch of the `if` statement.